Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/2-types/13-stable-types.md

---
sidebar_position: 13
---

> # Stable types

Reviewer: All of the examples here, though wrapped in actor, don't indent the their bodies. 
Unless @jessiemongeon1 instructed you otherwise, I would indent the bodies by two spaces.

| Mutable arrays (`[var T]`)                                                                   | Yes*   | No     |
| Records with immutable fields                                                                | Yes    | Yes    |
| Records with mutable fields                                                                  | Yes    | No     |
| Option types (`?T`)                                                                          | Yes*   | Yes**  |
| Variants with stable types                                                                   | Yes    | No (if non-shared types are included) |
| Shared functions                                                                             | Yes    | Yes    |
| Actor references                                                                             | Yes    | Yes    |
| Regions                                                                                      | Yes    | No     |
| Error type (`Error`)                                                                         | No     | No     |
| Non-shared functions and futures (`async* T`)                                                | No     | No     |

> \* provided `T` is stable  
> \** provided `T` is shared  

Reviewer 2: ```suggestion
\* provided `T` is stable.
\** provided `T` is shared.
```

:::info
In composite types like `[T]`, `[var T]`, and `?T`, the element type `T` must also be stable (for use in stable variables) or shared (for use in shared declarations). This ensures that the entire structure adheres to the requirements of stability or shareability, respectively.
Non-shared functions and futures (`async T`) and computations (`async* T`) depend on the current actor's code and execution context and therefore cannot be preserved across upgrades or transferred between actors.
:::

## Common stable types

### Primitive types

Most [primitive types](https://internetcomputer.org/docs/motoko/fundamentals/types/primitive-types) in Motoko are stable.

> ```motoko no-repl

Reviewer 2: The indentation for the examples on this page is incorrect

persistent actor {

> // Numbers, text, and booleans are implicitly stable
> // when an actor is declare with the persistent keyword

Reviewer: ```suggestion
// Numbers, text, Booleans and other primitive types are stable
```

persistent actor {
// Records with immutable fields are stable
var config = {
  appName = "My_Motoko_App";
  version = "1.0.0";
};

// Records with mutable fields are also stable
var settings = {
  var darkMode = false;
  var notifications = true;
  var port = 80;
  };
};
```

### Variants with stable type tags

[Variants](https://internetcomputer.org/docs/motoko/fundamentals/types/variants) are stable when their tags contain only stable types.

```motoko no-repl
persistent actor {
// Variants with stable tags are stable
type UserStatus = {
    #online;
    #offline;
    #busy : Text;
};

var motokoStatus : UserStatus = #online;
var ghostStatus : UserStatus = #busy("In a meeting");
};
```

### Option types

[Option](https://internetcomputer.org/docs/motoko/fundamentals/types/options) types are stable when they contain stable types.

```motoko no-repl
persistent actor {

// Option types with stable inner types are stable
var optionalDeadline : ?Nat = ?1640995200000;
var optionalMessage : ?Text = null;
};
```

### Regions

The [`Region`](https://internetcomputer.org/docs/motoko/base/Region) type, which provides low-level memory management, is stable.

```motoko no-repl

> 

Reviewer 2: ```suggestion
```

persistent actor {

// Regions are stable
var storage : Region = Region.new();
}
```

### Actor references

References to [actors](https://internetcomputer.org/docs/motoko/fundamentals/actors-async) are stable, allowing stable canister-to-canister interactions.

```motoko no-repl
persistent actor {
// Actor types are stable
type LoggerActor = actor {
    log : shared (message : Text) -> async ();
};
var logger : ?LoggerActor = null;
};
```

### Objects with mutable fields

> Objects with mutable fields are stable but not shared.

Reviewer: ```suggestion
Simple objects with mutable fields (but no methods) are stable. Such simple objects are the same as records.
```

---
sidebar_position: 13
---

# Stable types

> Recall that [shared types](https://internetcomputer.org/docs/motoko/fundamentals/types/shared-types) can be transmitted to other actors and canisters as the arguments or results of shared functions.

Reviewer 2: ```suggestion

```

Reviewer 2: Unnecessary, and we shouldn't assume that they have directly come from anything that explained shared types, so saying "Recall that" isn't valuable

> **Stable types** comprise all shared types and represent the types of values that may be stored in the stable declarations of a Motoko actor.

Reviewer 2: ```suggestion
**Stable types** include all [shared types](link) and represent the kinds of values that can be stored in the `stable` declarations of a Motoko actor.
```

> Storing a value in a stable declaration ensures that the value persists across actor upgrades, enabling state preservation without relying on a file system or database

Reviewer 2: ```suggestion
Storing a value in a `stable` declaration ensures that it persists across canister upgrades. This enables state preservation without the need for an external file system or database.
```

> The ability to retain data across upgrades is a key feature of Motoko, distinguishing it from other mainstream languages.

Reviewer 2: I would remove this, it isn't a very valuable comparison IMO

Reviewer: > I would remove this, it isn't a very valuable comparison IMO
Isn't that Motoko's USP though?

Reviewer 2: Sure, but in this context it doesn't really provide much value since 'upgrading' in the sense that we use it on ICP isn't the same as 'upgrading' a normal web app, so the comparison there is a little off.

Reviewer: I'm not sure I see the distinction but go ahead and delete the sentence.

> The set of stable types corresponds to the types of values that can be transferred from an actor to its future upgraded versions.

Reviewer 2: ```suggestion
The set of stable types defines the kinds of values that can be transferred from an actor to its future upgraded versions.
```

> Types that cannot be transferred include those whose values depend on the actor’s current code, such as non-shared functions or, more generally, objects with function members.

Reviewer 2: ```suggestion
Types that cannot be transferred include those whose values depend on the actor's current code, such as non-shared functions or, more generally, objects containing function members. These types are not stable because their behavior cannot be preserved independently of the code that defines them.
```

:::info  

> In actors declared without the `persistent` keyword, all private declarations are treated as transient unless explicitly marked `stable`.  
> In a `persistent` actor, all private declarations—except for function declarations—are considered stable by default, unless explicitly marked `transient`.  
> The type of any stable variable must belong to the set of stable types.  
> Transient variables are exempt from this restriction and may have any type, including non-stable types.  
> :::
> 

Reviewer 2: ```suggestion
In Motoko, the treatment of private declarations depends on whether an actor is declared with the `persistent` keyword:

- In actors **without** the `persistent` keyword, all private declarations are considered *transient* by default, unless explicitly marked `stable`.

- In **`persistent` actors**, all private declarations (except function declarations) are considered **stable** by default, unless explicitly marked `transient`.

Stable variables must have types that belong to the set of stable types.  
Transient variables are not subject to this restriction and may have any type, including non-stable types such as functions or objects with function members.

:::

```

| Records with immutable fields                                                                | Yes    | Yes    |
| Records with mutable fields                                                                  | Yes    | No     |
| Option types (`?T`)                                                                          | Yes*   | Yes**  |
| Variants with stable types                                                                   | Yes    | No (if non-shared types are included) |
| Shared functions                                                                             | Yes    | Yes    |
| Actor references                                                                             | Yes    | Yes    |
| Regions                                                                                      | Yes    | No     |
| Error type (`Error`)                                                                         | No     | No     |
| Non-shared functions and futures (`async* T`)                                                | No     | No     |

\* provided `T` is stable  
\** provided `T` is shared  

:::info

> In types such as `[T]`, `[var T]`, and `?T`, the element type `T` must itself be stable (for stable use) or shared (for shared use).  

Reviewer 2: ```suggestion
In composite types like `[T]`, `[var T]`, and `?T`, the element type `T` must also be stable (for use in stable variables) or shared (for use in shared declarations). This ensures that the entire structure adheres to the requirements of stability or shareability, respectively.
```

> Non-shared functions and futures (`async* T`) depend on the current actor’s execution context and cannot be preserved or transferred.  

Reviewer 2: ```suggestion
Non-shared functions and futures (`async* T`) depend on the current actor's execution context and therefore cannot be preserved across upgrades or transferred between actors.
```

Reviewer: (Jessie, I tweaked the suggestion a little).
